src/main/java/io/nats/client/support/WebSocket.java

```diff
@@ -0,0 +1,365 @@
+        }
+        // 1. expect 101:
+        if (!responseLine.toLowerCase().startsWith(WEBSOCKET_RESPONSE_LINE.toLowerCase())) {
+            throw new IllegalStateException("Expected " + WEBSOCKET_RESPONSE_LINE + ", but got " + responseLine);
+        }
+        Map<String, String> headers = new HashMap<>();
+        while (true) {
+            String line = readLine(buffer, in);
+            if (null == line) {
+                throw new IllegalStateException("Expected HTTP header to not exceed " + MAX_LINE_LEN);
+            }
+            if ("".equals(line)) {
+                break;
+            }
+            int colon = line.indexOf(':');
+            if (colon >= 0) {
+                if (headers.size() >= MAX_HTTP_HEADERS) {
+                    throw new IllegalStateException("Exceeded max HTTP headers=" + MAX_HTTP_HEADERS);
+                }
+                headers.put(
+                    line.substring(0, colon).trim().toLowerCase(),
+                    line.substring(colon + 1).trim());
+            } else {
+                throw new IllegalStateException("Expected HTTP header to contain ':', but got " + line);
+            }
+        }
+        // 2. Expect `Upgrade: websocket`
+        if (!"websocket".equalsIgnoreCase(headers.get("upgrade"))) {
+            throw new IllegalStateException(
+                "Expected HTTP `Upgrade: websocket` header");
+        }
+        // 3. Expect `Connection: Upgrade`
+        if (!"upgrade".equalsIgnoreCase(headers.get("connection"))) {
+            throw new IllegalStateException(
+                "Expected HTTP `Connection: Upgrade` header");
+        }
+        // 4. Sec-WebSocket-Accept: base64(sha1(key + "258EAF..."))
+        MessageDigest sha1;
+        try {
+            sha1 = MessageDigest.getInstance("SHA-1");
+        } catch (NoSuchAlgorithmException ex) {
+            throw new IllegalStateException(ex);
+        }
+        sha1.update(key.getBytes(UTF_8));
+        sha1.update("258EAFA5-E914-47DA-95CA-C5AB0DC85B11".getBytes(UTF_8));
+        String acceptKey = Base64.getEncoder().encodeToString(
+            sha1.digest());
+        String gotAcceptKey = headers.get("sec-websocket-accept");
+        if (!acceptKey.equals(gotAcceptKey)) {
+            throw new IllegalStateException(
+                "Expected HTTP `Sec-WebSocket-Accept: " + acceptKey + ", but got " + gotAcceptKey);
+        }
+        // 5 & 6 are not valid, since nats-server doesn't
+        // implement extensions or protocols.
+    }
+
+    private static String readLine(byte[] buffer, InputStream in) throws IOException {
+        int offset = 0;
+        int lastCh = -1;
+        while (true) {
+            int ch = in.read();
+            switch (ch) {
+            case -1:
+                // Premature EOF (everything should be terminated with \n)
+                return new String(buffer, 0, offset);
```

Comment:
I forgot to explicitly specify the charset here and 3 lines down. This should be:

```
return new String(buffer, 0, offset, StandardCharsets.ISO_8859_1);
```

...the choice of ISO_8859_1 is purely because every byte maps to a character and because this is a super-set of ascii.

The only issue with the existing code is if the default JVM character set is not compatible with ASCII (such as an EBCDIC variant).

I realize this is asymmetric with the fact that the HTTP request uses UTF-8, but the only code using the response from the `readLine()` is only processing very specific headers for which ASCII is all we care about.

Reply:
Fixed